integration-libs/punchout/root/services/punchout-state-persistence.service.ts

```diff
@@ -20,51 +23,68 @@ export class PunchoutStatePersistanceService implements OnDestroy {
   protected subscription = new Subscription();
   protected hasPunchoutStarted = false;
-
   /**
    * Initializes the synchronization between state and browser storage.
-   * Through getPunchoutSessionId(), storage is updated everytime PunchoutState is modified.
+   * Through getPunchoutStorageInfo(), storage is updated everytime PunchoutState is modified.
    */
   public initSync() {
     this.subscription.add(
       this.statePersistenceService.syncWithStorage({
         key: PUNCHOUT_STORAGE_KEY,
-        state$: this.getPunchoutSessionId(),
+        state$: this.getPunchoutStorageInfo(),
         onRead: (state) => this.onRead(state),
       })
     );
   }
-
   /**
    * Gets and transforms state into the form that should
    * be saved in storage.
    */
-  protected getPunchoutSessionId(): Observable<string | undefined> {
+  protected getPunchoutStorageInfo(): Observable<
+    punchoutStorageInfo | undefined
+  > {
     return this.punchoutStoreService.getPunchoutState().pipe(
       map((punchoutState) => {
-        if (punchoutState?.punchoutSessionId) {
+        if (
+          punchoutState?.punchoutSessionId &&
+          punchoutState?.punchoutSession?.punchOutOperation
+        ) {
           this.hasPunchoutStarted = true;
-          return punchoutState?.punchoutSessionId;
+          return {
+            punchoutSessionId: punchoutState?.punchoutSessionId,
+            punchOutOperation: punchoutState.punchoutSession.punchOutOperation,
+          };
         }
         // With 'undefined' value, no key/value gets modified or created, it keeps the storage cleaned when Punchout is unused.
         // Note that StatePersistenceService does not allow to delete key/value once it has been created.
-        return this.hasPunchoutStarted ? '' : undefined;
+        return this.hasPunchoutStarted
+          ? { punchoutSessionId: '', punchOutOperation: '' }
+          : undefined;
       })
     );
   }
-
   /**
    * Function called on each browser storage read.
    * Used to update state from browser -> state.
-   * storage stores minimum data: only punchoutSessionId.
+   * storage stores minimum data: punchoutSessionId and punchOutOperation.
    * Full PunchoutSession object is retrieved by calling punchoutFacade.getPunchoutSession
    */
-  protected onRead(punchoutSessionId: string | undefined) {
+  protected onRead(punchoutStorageInfo: punchoutStorageInfo | undefined) {
+    if (!punchoutStorageInfo) {
+      return;
+    }
+    const { punchoutSessionId, punchOutOperation } = punchoutStorageInfo;
     if (
       punchoutSessionId &&
+      punchOutOperation &&
       !this.punchoutDetectionService.isPunchoutSessionPage()
     ) {
-      this.punchoutStoreService.setPunchoutState({ punchoutSessionId });
+      this.punchoutStoreService.setPunchoutState({
+        punchoutSessionId,
+        punchoutSession: {
+          punchOutOperation: punchOutOperation as PunchOutOperation,
+        },
```

Comment:
[QUESTION]
Why we need store punchout operation in local storage? Is it really related to adding an Angular Guard? Or it was needed anyway and we just added it in this PR by the way?

Reply:
to get the punchoutOperation value, we have to call the punchout Session Api and wait for its response.
It was for performance matter.